Add Mask Morphology plugin with erosion and dilation operations

decode-soumyadipta opened the pull request:
# Mask Morphology Plugin

This plugin adds morphological operations (erosion and dilation) to InVesalius, allowing users to refine their masks by removing small artifacts or filling gaps.

## Features

- Apply erosion or dilation to the current mask
- Choose between 2D (slice-by-slice) or 3D (volumetric) operations
- Adjust the radius (1-10) to control the strength of the operation
- Detailed logging to track progress and results

## Implementation Details

- Uses scikit-image's binary_erosion and binary_dilation functions
- Supports both disk (2D) and ball (3D) structuring elements
- Properly handles InVesalius mask data structure
- Simple and intuitive user interface

## Files Added

- `mask_morphology/__init__.py` - Package initialization
- `mask_morphology/plugin.json` - Plugin metadata
- `mask_morphology/main.py` - Entry point for the plugin
- `mask_morphology/gui.py` - Main implementation with UI and processing logic

I've tested this with various masks and it works well for cleaning up segmentations.

tfmoraes commented on mask_morphology/gui.py:
```diff
@@ -0,0 +1,171 @@
+            mask_matrix[1:, 1:, 1:] = result.astype(np.uint8) * 255
+            
+            # Update the mask in the project
+            current_mask.matrix = mask_matrix
```
add:

```
current_mask.was_edited = True
```

tfmoraes commented on mask_morphology/main.py:
```diff
@@ -0,0 +1,12 @@
+    """
+    top_window = wx.GetApp().GetTopWindow()
+    window = gui.MaskMorphologyGUI(top_window)
+    window.ShowModal()
```
Use `window.Show()` so user can interact with slices to see what is happening.

decode-soumyadipta replied:
Done sir

tfmoraes commented on mask_morphology/gui.py:
```diff
@@ -0,0 +1,171 @@
+            logger.info("Visualization updated")
+            
+            # Show success message
+            wx.MessageBox(f"{op_name} with radius {radius} applied successfully.", 
```
I don't think it's needed to show a message.

decode-soumyadipta replied:
Yes , i completely agree with you.

decode-soumyadipta commented:
Hi Thiago,

I've made all the requested changes to the Mask Morphology plugin:

1. Added `current_mask.was_edited = True` in gui.py to properly mark the mask as edited
2. Removed the success message box as it's not needed
3. Changed `window.ShowModal()` to `window.Show()` in main.py to allow users to interact with slices while the plugin is open

The changes have been pushed to the same branch. Could you please review the updated PR when you have a chance?

Thank you for your guidance!

Best regards,
Soumyadipta

tfmoraes commented on mask_morphology/gui.py:
```diff
@@ -0,0 +1,175 @@
+                             "Empty Result", wx.OK | wx.ICON_WARNING)
+            
+            # Update the mask
+            mask_matrix[1:, 1:, 1:] = result.astype(np.uint8) * 255
```
Sorry, I forgot to say. You have to do this too:

```
mask_matrix[0] = 1
mask_matrix[:, 0, :] = 1
mask_matrix[:, :, 0] = 1
```

decode-soumyadipta replied:
I've made all the requested changes to the Mask Morphology plugin:

1. Added current_mask.was_edited = True in gui.py to properly mark the mask as edited
2. Removed the success message box as it's not needed
3. Changed window.ShowModal() to window.Show() in main.py to allow users to interact with slices while the plugin is open

decode-soumyadipta commented:
I've updated the plugin based on your feedback. The changes include:

1. Fixed the mask border handling by setting the first slice, row, and column to 1 as you suggested
2. Improved the result handling to prevent empty masks
3. Added better error checking and logging
4. Made sure the mask is properly updated with the morphological operation results

The plugin now correctly applies both erosion and dilation operations while maintaining the proper In

decode-soumyadipta commented:
I've implemented the changes you suggested in your comment about setting the first slice, row, and column to 1 in the mask matrix. However, I encountered some issues during the commit process.

When pushing my changes, `I accidentally included some unrelated files in the PR. I've since tried to clean up the branch and delete them, but you might still see these extra files in the "Files changed" section of the PR.` Please note that only the files under the mask_morphology directory are relevant to my contribution:

- mask_morphology/gui.py
- mask_morphology/main.py
- mask_morphology/__init__.py
- mask_morphology/plugin.json

I would greatly appreciate if you could focus your review on just these files. The core functionality is working correctly now - both erosion and dilation operations maintain the proper InVesalius mask format with the border values set to 1 as you suggested.

Thank you for your understanding and guidance throughout this process. I'm learning a lot about contributing to open source projects.

tfmoraes commented:
Try to clean it up and fix this PR or create a new one.

decode-soumyadipta commented:
> Try to clean it up and fix this PR or create a new one.

@tfmoraes 

Finally, I've now implemented all the changes you suggested and thoroughly tested the plugin. 


The files are now cleanly organized in the proper directory structure in this PR itself. The mask morphology plugin works correctly for both erosion and dilation operations. 

Please review the final implementation and let me know if any further adjustments are needed. 

Thank you for your guidance😊! Blessed to have a mentor like you who brings out my best technical potential. Appreciating your patience!

![lnl](https://github.com/user-attachments/assets/b967fba9-fd06-4575-a215-ce72b7497d10)

Terminal-

![,j,b](https://github.com/user-attachments/assets/3ac46d70-5642-4e00-80d4-3a8e92d3683f)

decode-soumyadipta commented:
I made all the requested changes and made the PR, clean & well structured.
Please review it and let me know what else I can do!

Thanks,
Soumyadipta Dey

On Sat, Mar 15, 2025 at 1:16 AM Thiago Franco de Moraes <
***@***.***> wrote:

> Try to clean it up and fix this PR or create a new one.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/invesalius/inv3_plugins/pull/4#issuecomment-2725607611>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDR7RQWUV425X33LZNW5RSL2UMWZLAVCNFSM6AAAAABY5FLRFWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRVGYYDONRRGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: tfmoraes]*tfmoraes* left a comment (invesalius/inv3_plugins#4)
> <https://github.com/invesalius/inv3_plugins/pull/4#issuecomment-2725607611>
>
> Try to clean it up and fix this PR or create a new one.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/invesalius/inv3_plugins/pull/4#issuecomment-2725607611>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDR7RQWUV425X33LZNW5RSL2UMWZLAVCNFSM6AAAAABY5FLRFWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRVGYYDONRRGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

decode-soumyadipta commented:
@tfmoraes 

Could you please review it sir?

tfmoraes commented:
thanks @decode-soumyadipta!

decode-soumyadipta commented:
Thank you for the guidance and support sir. Looking forward for your support in GSOC 2025. 

Support me contributing! 😊
